### DISCUSSION TITLE
stream url_mp3_helix_i2s no audio output at MAX98357A

### DISCUSSION THREAD

**Discussion by wuppergit** [pschatzmann/arduino-audio-tools#1359] [Q&A] [answered]
I'm struggling to get audio output to MAX98357A from url. There is only one MAX98357A .No sound, no noise, nothing.
The basis is the example-sketch stream url_mp3_helix_i2s IDE Arduino 1.8.19, Espressif 2.0.14, ESP32). The output on the serial monitor is like this:

> [I] AudioTypes.h : 104 - sample_rate: 44100
[I] AudioTypes.h : 105 - channels: 2
[I] AudioTypes.h : 106 - bits_per_sample: 16
[I] I2SConfigESP32.h : 90 - rx/tx mode: TX_MODE
[I] I2SConfigESP32.h : 91 - port_no: 0
[I] I2SConfigESP32.h : 92 - is_master: Master
[I] I2SConfigESP32.h : 93 - sample rate: 44100
[I] I2SConfigESP32.h : 94 - bits per sample: 16
[I] I2SConfigESP32.h : 95 - number of channels: 2
[I] I2SConfigESP32.h : 96 - signal_type: Digital
[I] I2SConfigESP32.h : 98 - i2s_format: I2S_STD_FORMAT
[I] I2SConfigESP32.h : 100 - auto_clear: true
[I] I2SConfigESP32.h : 102 - use_apll: true
[I] I2SConfigESP32.h : 107 - buffer_count:6
[I] I2SConfigESP32.h : 108 - buffer_size:512
[I] I2SConfigESP32.h : 113 - pin_bck: 25
[I] I2SConfigESP32.h : 115 - pin_ws: 26
[I] I2SConfigESP32.h : 117 - pin_data: 27
[I] AudioEncoded.h : 269 - virtual void audio_tools::EncodedAudioOutput::setNotifyAudioChange(audio_tools::AudioInfoSupport&)
[I] URLStream.h : 85 - virtual bool audio_tools::URLStream::begin(const char*, const char*, MethodID, const char*, const char*): http://stream.srg-ssr.ch/m/rsj/mp3_128
[I] Url.h : 74 - Url::parse
[I] Url.h : 112 - url->http://stream.srg-ssr.ch/m/rsj/mp3_128
[I] Url.h : 113 - host->stream.srg-ssr.ch
[I] Url.h : 114 - protocol->http
[I] Url.h : 115 - path->/m/rsj/mp3_128
[I] Url.h : 116 - port->80
[I] URLStream.h : 382 - void audio_tools::URLStream::login()
.
[I] URLStream.h : 356 - WiFiClient
[I] HttpRequest.h : 231 - process connecting to host stream.srg-ssr.ch port 80
[I] HttpRequest.h : 329 - connected 1 timeout 60000
[I] HttpRequest.h : 242 - Free heap: 183516
[I] HttpHeader.h : 284 - HttpHeader::write
[I] HttpHeader.h : 430 - -> GET /m/rsj/mp3_128 HTTP/1.1
[I] HttpHeader.h : 206 -  -> Host: stream.srg-ssr.ch 
[I] HttpHeader.h : 206 -  -> Connection: keep-alive 
[I] HttpHeader.h : 206 -  -> Accept-Encoding: identity 
[I] HttpHeader.h : 206 -  -> Accept: audio/mp3 
[I] HttpHeader.h : 349 -  -> <CR LF> 
[I] HttpRequest.h : 283 - Request written ... waiting for reply
[W] URLStream.h : 323 - Redirected to: http://fritz.box/tools/kids_not_allowed.lua?account=landeviceUID-9381
[I] Url.h : 74 - Url::parse
[I] Url.h : 112 - url->http://fritz.box/tools/kids_not_allowed.lua?account=landeviceUID-9381
[I] Url.h : 113 - host->fritz.box
[I] Url.h : 114 - protocol->http
[I] Url.h : 115 - path->/tools/kids_not_allowed.lua?account=landeviceUID-9381
[I] Url.h : 116 - port->80
[I] HttpRequest.h : 231 - process connecting to host fritz.box port 80
[I] HttpRequest.h : 329 - connected 1 timeout 60000
[I] HttpRequest.h : 242 - Free heap: 182424
[I] HttpHeader.h : 284 - HttpHeader::write
[I] HttpHeader.h : 430 - -> GET /tools/kids_not_allowed.lua?account=landeviceUID-9381 HTTP/1.1
[I] HttpHeader.h : 206 -  -> Host: fritz.box 
[I] HttpHeader.h : 206 -  -> Connection: keep-alive 
[I] HttpHeader.h : 206 -  -> Accept-Encoding: identity 
[I] HttpHeader.h : 206 -  -> Accept: audio/mp3 
[I] HttpHeader.h : 349 -  -> <CR LF> 
[I] HttpRequest.h : 283 - Request written ... waiting for reply
[I] HttpHeader.h : 252 - Waiting for data...
[I] HttpHeader.h : 261 - Data available: 5744
[I] HttpRequest.h : 173 - no CONTENT_LENGTH found in reply
[I] URLStream.h : 110 - size: 0
[I] URLStream.h : 117 - ==> http status: 200
[I] StreamCopy.h : 143 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 143 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 143 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 143 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 143 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 143 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 143 - StreamCopy::copy  368 -> 368 -> 368 bytes - in 1 hops

So in my opinion, there are some incoming bytes and they are moved to out. I2S and wiring are working (no issues with streams-memory_mp3_short-i2s, player-spiffs-i2s). URL is working in chrome-browser too.
What am I doing wrong.

**Comment by pschatzmann** [comment]
I would suggest to simplify the setup and test with a sine generator first.
Most likely you did not connect the correct pins or your module is defect... 

If you have an Oscilloscope you could double check the signal on the pins.

**Reply by wuppergit** [reply]
I tried witth streams-memory_mp3_short-i2s- and  player-spiffs-i2s-example. Both are working fine. So I don't think the issue is mistake in  wiring or pin-assignment. 
I'm not so well equiped to have an oscilliscope

**Reply by pschatzmann** [reply]
Hmm, the only other explanation that I would have is, that you are not feeding any valid mp3 data...

**Reply by wuppergit** [reply]
Is that the reason for the log posts copied bytes? And additionaly, a call in chrome browser without the parameter "audio/mp3" is working well?

**Reply by pschatzmann** [reply]
If you consistently receive the copy log entries, then it means that the source is feeding the destination.
You can try do output the data as CSV on the Serial Plotter, just to double check the result of the decoding...

**Reply by wuppergit** [reply]
Good morning,
in the recent hours I tried a lot (after installing fresh audiotools and libhelix):
- output data as CSV on the Serial Plotter (streams-url-raw-serial)
  There is output - lots of value-pairs. Output stops after 7 hops (like log in first post); but why does it stop?
- output data as analog
  example streams-memory_mp3-analog is working fine; streams-url_mp3-analog not (stop after 7 hops, no audio)
- example streams-tts-i2s compile error; "TTS.h" no such file
- example streams-url-aac-i2s; no audio / hops with the specified url

So I stop working now. 
Thank you very much anyway, still great work.

**Reply by pschatzmann** [reply]
If you have only 7 copy lines, then the issue is with your url and server which stops to send data. Since you get the data from a local server, did you check in it's log why it is stopping to send data ?


In any case you need to be more systematic:
https://github.com/pschatzmann/arduino-audio-tools/wiki/It's-not-working

- Did you run the sine test to check if your DAC is working ?
- The output of value pairs, does not mean much: you need to look at it in the plotter to see if you get any sound waves!
- Is your file really only 6k big?
- Maybe it is an authentication problem ? Is http://fritz.box/tools/kids_not_allowed.lua?account=landeviceUID-9381 really the right final address ?
- The Browser might do some fancy things: did you test with curl ?
- fritz.box should support ftp, you could try to stream the data via ftp
- Finally it is pointless to try to even add more complexity with the TTS example which requires an additional library.

**Reply by wuppergit** [reply]
- sine-wave-example is working fine
- plotter prints waves but stops (example stream-url-raw-serial)
- I don't know where the sketch takes the input of 6k from
- I don't know, what fritzbox is doing internally - sorry
- Trying ` Invoke-WebRequest -URI http://stream.srg-ssr.ch/m/rsj/mp3_128` is working fine - lots of bytes receiving
- fritzbox ftp / RPi-server may by tasks for next weekend

**Reply by wuppergit** [reply, marked as answer]
Hello,
I figured it out now.
The cause lies in the fritzbox, especially in the parental controls. It was activated for the ESP32; no idea when, why, by whom and how. 
The leading hint was `URLStream.h : 323 - Redirected to: [http://fritz.box/tools/kids_not_allowed.lua](http://fritz.box/tools/kids_not_allowed.lua?account=landeviceUID-9381)`
Everything is fine now
Thank you again

**Comment by pschatzmann** [comment]
So the issue is on the server side!

Here is the [ftp example](https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-communication/ftp/ftp-client/ftp-client.ino)
